src/index.js

```diff
@@ -1,390 +1,586 @@
+}
 
-        var oldKey = getKey(oldChildren[i])
-        if (oldKey != null) {
-          oldKeyed[oldKey] = [oldElements[i], oldChildren[i]]
-        }
+var update = function(sub, oldSub, dispatch) {
+  if (sub.effect === oldSub.effect) {
+    for (var k in merge(sub.props, oldSub.props)) {
+      if (sub.props[k] !== oldSub.props[k]) {
+        cancel(oldSub)
+        return start(sub, dispatch)
       }
+    }
+  }
+  return oldSub
+}
 
-      var i = 0
-      var k = 0
+var start = function(sub, dispatch) {
+  return merge(sub, {
+    cancel: sub.effect(sub.props, dispatch)
+  })
+}
 
-      while (k < children.length) {
-        var oldKey = getKey(oldChildren[i])
-        var newKey = getKey((children[k] = resolveNode(children[k])))
+var listen = function(sub, oldSub, dispatch) {
+  if (isArray(sub) || isArray(oldSub)) {
+    var out = []
+    var subs = isArray(sub) ? sub : [subs]
+    var oldSubs = isArray(oldSub) ? oldSub : [oldSubs]
 
-        if (newKeyed[oldKey]) {
-          i++
-          continue
-        }
+    for (var i = 0; i < subs.length || i < oldSubs.length; i++) {
+      out.push(listen(subs[i], oldSubs[i], dispatch))
+    }
 
-        if (newKey != null && newKey === getKey(oldChildren[i + 1])) {
-          if (oldKey == null) {
-            removeElement(element, oldElements[i], oldChildren[i])
-          }
-          i++
-          continue
-        }
+    return out
+  }
 
-        if (newKey == null || isRecycling) {
-          if (oldKey == null) {
-            patch(element, oldElements[i], oldChildren[i], children[k], isSvg)
-            k++
-          }
-          i++
-        } else {
-          var keyedNode = oldKeyed[newKey] || []
+  return sub
+    ? oldSub ? update(sub, oldSub, dispatch) : start(sub, dispatch)
+    : oldSub ? cancel(oldSub) : oldSub
+}
+/* OR
+var toArray = function(any) {
+  return isArray(any) ? any : any == null ? [] : [any]
+}
 
-          if (oldKey === newKey) {
-            patch(element, keyedNode[0], keyedNode[1], children[k], isSvg)
-            i++
-          } else if (keyedNode[0]) {
-            patch(
-              element,
-              element.insertBefore(keyedNode[0], oldElements[i]),
-              keyedNode[1],
-              children[k],
-              isSvg
-            )
-          } else {
-            patch(element, oldElements[i], null, children[k], isSvg)
-          }
+var batch = function(subs, oldSubs, dispatch) {
+  var out = []
+  for (var i = 0; i < subs.length || i < oldSubs.length; i++) {
+    out.push(listen(subs[i], oldSubs[i], dispatch))
+  }
+  return out
+}
 
-          newKeyed[newKey] = children[k]
-          k++
-        }
+export var listen = function(sub, oldSub, dispatch) {
+  return isArray(sub) || isArray(oldSub)
+    ? batch(toArray(sub), toArray(oldSub), dispatch)
+    : sub
+      ? oldSub ? update(sub, oldSub, dispatch) : start(sub, dispatch)
+      : oldSub ? cancel(oldSub) : oldSub
+}
+*/
+
+export function app(props) {
```

Comment:
How about instead of inline exports, using the following block at the top of the bottom of the file
```jsx
export { h, app }
```
To make clear that `h` and `app` will be exported.

I am using this pattern and I found it verry clever. (I learned it from Erlang).

Reply:
I _could_ like that. Let me do some research.

Reply:
It is definitely more clear when all exports goes near to each other, but I found out that for example webpack generate more noise when you export a name instead of an expression.